src/plugins/controls/public/control_group/embeddable/control_group_container.tsx

```diff
@@ -31,117 +40,277 @@ import { DEFAULT_CONTROL_WIDTH } from '../editor/editor_constants';
+     */
+    this.subscriptions.add(
+      this.getInput$()
+        .pipe(
+          skip(1),
+          distinctUntilChanged((a, b) => controlOrdersAreEqual(a.panels, b.panels))
+        )
+        .subscribe(() => {
+          this.recalculateDataViews();
+          this.recalculateFilters();
+          this.childOrderCache = this.getEmbeddableOrderCache();
+          this.childOrderCache.idsInOrder.forEach((id) =>
+            this.getChild(id)?.refreshInputFromParent()
+          );
+        })
+    );
+
+    /**
+     * Create a pipe that outputs the child's ID, any time any child's output changes.
+     */
+    const anyChildChangePipe = pipe(
+      map(() => this.getChildIds()),
+      distinctUntilChanged(deepEqual),
+
+      // children may change, so make sure we subscribe/unsubscribe with switchMap
+      switchMap((newChildIds: string[]) =>
+        merge(
+          ...newChildIds.map((childId) =>
+            this.getChild(childId)
+              .getOutput$()
+              .pipe(
                 // Embeddables often throw errors into their output streams.
-                .pipe(catchError(() => EMPTY))
-            )
+                catchError(() => EMPTY),
+                mapTo(childId)
+              )
           )
         )
-      );
+      )
+    );
 
-      this.subscriptions.add(
-        merge(this.getOutput$(), this.getOutput$().pipe(anyChildChangePipe))
-          .pipe(debounceTime(10))
-          .subscribe(this.recalculateOutput)
-      );
+    /**
+     * run OnChildOutputChanged when any child's output has changed
+     */
+    this.subscriptions.add(
+      this.getOutput$()
+        .pipe(anyChildChangePipe)
+        .subscribe((childOutputChangedId) => {
+          this.recalculateDataViews();
+          if (childOutputChangedId === this.childOrderCache.lastChildId) {
+            // the last control's output has updated, recalculate filters
+            this.recalculateFilters$.next();
+            return;
+          }
+
+          // when output changes on a child which isn't the last - make the next embeddable updateInputFromParent
+          const nextOrder = this.childOrderCache.IdsToOrder[childOutputChangedId] + 1;
+          if (nextOrder >= Object.keys(this.children).length) return;
+          setTimeout(
+            () =>
+              this.getChild(this.childOrderCache.idsInOrder[nextOrder]).refreshInputFromParent(),
+            1 // run on next tick
+          );
+        })
+    );
+
+    /**
+     * debounce output recalculation
+     */
+    this.subscriptions.add(
+      this.recalculateFilters$.pipe(debounceTime(10)).subscribe(() => this.recalculateFilters())
+    );
+  };
+
+  private initializeChildrenInOrder = async () => {
+    const { panels } = this.getInput();
+    for (const childId of this.childOrderCache.idsInOrder) {
+      await this.createEmbeddableFromPanel(panels[childId]);
+      await this.untilEmbeddableLoaded(childId);
+    }
+  };
```

Comment:
I know this is a very specific use case, but I think we should move this functionality to the abstract `Container` class. It seems like the API could be a bit clearer if we do so:

```ts
constructor(
    input: TContainerInput,
    output: TContainerOutput,
    protected readonly getFactory: EmbeddableStart['getEmbeddableFactory'],
    parent?: IContainer,
    options?: { initChildrenById: string[] } // an options object would also be easier to refactor if we ever have more.
) {
}
```

Rather than passing a boolean that skips the first round of builds. From the abstract class perspective the reasoning is a bit opaque if we just pass a flag IMO. WDYT @vadimkibana ? We can also add some docs:

```
/**
 * Initialise children in the order specified. If an ID does not match it will be skipped and if a child is not included it will be initialized in the default order after the list of provided IDs.
 */
initChildrenById: string[]
```

Reply:
either way a test would be 👍🏻 👍🏻

Reply:
I did it this way in an attempt to minimize the changes to `container.ts`, and because the order is stored in the panels object in a way that is inaccessible to the generic `container.ts`. But I can see how the flag creates an API which isn't well defined.

I do like your suggestion better - but the args object may need two parameters instead of just one to deal with another difference required by the hierarchical chain, so it might look like:

`childIdInitializeOrder`: An ordered array of IDs 
`waitUntilEachChildReady`: If true after initializing an embeddable, the container will wait for it to finish loading before initializing the next embeddable.

This could also potentially add more flexibility to the design. It is a pretty specific use-case, but I can imagine other containers having some requirement for initializing children in order.

I'm curious to hear @vadimkibana's thoughts on the matter. Either way, I will definitely add a test. Thanks for the suggestions @jloleysens !